pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go

```diff
@@ -0,0 +1,101 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package dhcp
+
+import (
+	"bytes"
+	"net"
+	"time"
+
+	dhcp "github.com/krolaw/dhcp4"
+	dhcpConn "github.com/krolaw/dhcp4/conn"
+
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const infiniteLease = 999 * 24 * time.Hour
+
+func SingleClientDHCPServer(
+	clientMAC net.HardwareAddr,
+	clientIP net.IP,
+	clientMask net.IPMask,
+	serverIface string,
+	serverIP net.IP,
+	routerIP net.IP,
+	dnsIP net.IP) {
+
+	log.Log.Info("Starting SingleClientDHCPServer")
+
+	handler := &DHCPHandler{
+		clientIP:      clientIP,
+		clientMAC:     clientMAC,
+		serverIP:      serverIP,
+		leaseDuration: infiniteLease,
+		options: dhcp.Options{
+			dhcp.OptionSubnetMask:       []byte(clientMask),
+			dhcp.OptionRouter:           []byte(routerIP),
+			dhcp.OptionDomainNameServer: []byte(dnsIP),
+		},
+	}
+
+	l, err := dhcpConn.NewUDP4BoundListener(serverIface, ":67")
+	if err != nil {
+		panic(err)
```

Comment:
Can you just use an error channel?

Reply:
what would we do with the error channel in this case?

There's an entrypoint.sh script that's the actual pid 1 of the container.  I'm going to add logic in that script to send SIGTERM and eventually SIGKILL to any qemu processes that could still be running as a result of virt-launcher panicking. This means we'll still have a sane shutdown for qemu even if virt-launcher segfaults/panics/etc...

Reply:
I added #713 to address the orphaned qemu process.

Reply:
my main problem was the error handling in general (throwing, catching, trowing again). The channel is not so important.